oak-jcr/src/main/java/org/apache/jackrabbit/oak/jcr/api/binary/HttpBinaryProvider.java

```diff
@@ -0,0 +1,165 @@
+     * the upload again with a higher or correct size.
+     *
+     * <p>
+     * The caller needs to specify how many parts it can handle in {@code maxParts}. Each part
+     * will be a URL of some length, which typically need to be transported to a remote client,
+     * so there might be limitations on the overall number. Specifying a number too small might
+     * lead to an error, depending on the underlying provider limitations regarding minimum and
+     * maximum part size. The upper bound is dependent upon the service provider as well as the
+     * underlying implementation, but should be at least ten thousand.  Specifying a number too
+     * large may exceed implementation or service provider limitations and result in an error.
+     * Specifying -1 as the value of {@code maxParts} is supported, which means the caller can
+     * support any number of URLs being returned. The implementation will then choose the number
+     * of URLs to return, which could be as high as the implementation or service provider limitation.
+     *
+     * <p>
+     * The returned instructions will usually be time limited and cannot be shared with other users.
+     * They must only be returned to authenticated requests corresponding to this session user
+     * or trusted system components (service users).
+     *
+     * <p>
+     * The caller must provide the path to the node where the resulting binary is intended to be
+     * added. While calling this method itself should not result in any modification to the
+     * underlying store (that should happen later, in {@link #completeHttpUpload(String)},
+     * the path must be included in order to verify that the caller has permissions to add
+     * a binary property at the specified path. If the current session has not enough permissions
+     * to add binaries to the repository, an {@link AccessDeniedException} will be thrown.
+     *
+     * <p>
+     * This method does not affect the transient space of the current session.
+     *
+     * <p>
+     * If the upload fails or never happens before the URLs expire, no change will happen to
+     * the repository.
+     *
+     * @param path the path of the node to which the binary will be added
+     * @param maxSize the exact size of the binary to upload, if known,
+     *                or the maximum estimated size (required, must be larger than zero)
+     * @param maxParts maximum number of parts to return (required, must not be zero)
+     *
+     * @return HTTP upload instructions or {@code null} if the feature is not available
+     *
+     * @throws AccessDeniedException if the feature is available but the session is not allowed to add binaries,
+     * or RepositoryException if a more general repository error occurs
+     */
+    @Nullable
+    BinaryHttpUpload initializeHttpUpload(String path, long maxSize, int maxParts) throws RepositoryException;
```

Comment:
We should add a new checked exception for the known case that the client has provided `maxSize` & `maxParts` numbers that don't work, e.g. `UnsupportedHttpUploadArgumentsException` or sth similar.

IIUC, currently we would be returning `null`, but I think there is a difference between

* feature not supported at all by Oak setup/Datastore (for what returning `null` is meant for)
* feature supported, but client provided invalid values for arguments (new checked exception)

Reply:
We can do that in the lower level code.  The use of `performNullable()` in `SessionImpl.initializeHttpUpload()` can only return a `RepositoryException` though so unless we change that it will still end up being a null that is given back to the client.

Reply:
AFAICS, this is not a problem. SessionImpl throws only RepositoryException for many methods that per API throw others as well. So it might just be a question of adding them to the interface method and documenting it.

In any case, this is important for the client of the API here. If it doesn't just work, we need to find a way handle this inside SessionImpl.